src/EventStore.Core/TransactionLog/Scavenging/Stages/ChunkExecutor.cs

```diff
@@ -0,0 +1,375 @@
+								physicalWeight);
+						}
+						cancellationToken.ThrowIfCancellationRequested();
+					},
+					emitCheckpoint: chunkEndNumber => {
+						// this is called on the thread that called the loop, which does not do any of
+						// the processing.
+						// it is called after an item has been processed and before the slot is used
+						// to process another item. this gives us a meaningful opportunity to rest.
+						state.SetCheckpoint(
+							new ScavengeCheckpoint.ExecutingChunks(
+								scavengePoint,
+								chunkEndNumber));
+
+						if (_threads == 1) {
+							_throttle.Rest(cancellationToken);
+						} else {
+							// running a multithreaded scavenge with throttle < 100
+							// is rejected by the AdminController.
+						}
+					});
+			} finally {
+				for (var i = 0; i < borrowedStates.Length; i++) {
+					borrowedStates[i].Dispose();
+				}
+			}
+		}
+
+		private IEnumerable<IChunkReaderForExecutor<TStreamId, TRecord>> GetAllPhysicalChunks(
+			int startFromChunk,
+			ScavengePoint scavengePoint) {
+
+			var scavengePos = _chunkSize * startFromChunk;
+			var upTo = scavengePoint.Position;
+			while (scavengePos < upTo) {
+				// in bounds because we stop before the scavenge point
+				var physicalChunk = _chunkManager.GetChunkReaderFor(scavengePos);
+
+				if (!physicalChunk.IsReadOnly)
+					throw new Exception(
+						$"Reached open chunk before scavenge point. " +
+						$"Chunk {physicalChunk.ChunkStartNumber}. ScavengePoint: {upTo}.");
+
+				yield return physicalChunk;
+
+				scavengePos = physicalChunk.ChunkEndPosition;
+			}
+		}
+
+		private void ExecutePhysicalChunk(
+			float physicalWeight,
+			ScavengePoint scavengePoint,
+			IScavengeStateForChunkExecutorWorker<TStreamId> state,
+			ITFChunkScavengerLog scavengerLogger,
+			IChunkReaderForExecutor<TStreamId, TRecord> sourceChunk,
+			Stopwatch sw,
+			CancellationToken cancellationToken) {
+
+			sw.Restart();
+
+			int chunkStartNumber = sourceChunk.ChunkStartNumber;
+			long chunkStartPos = sourceChunk.ChunkStartPosition;
+			int chunkEndNumber = sourceChunk.ChunkEndNumber;
+			long chunkEndPos = sourceChunk.ChunkEndPosition;
+			var oldChunkName = sourceChunk.Name;
+
+			Log.Trace(
+				"SCAVENGING: started to scavenge physical chunk: {oldChunkName} " +
+				"with weight {physicalWeight:N0}. " +
+				"{chunkStartNumber} => {chunkEndNumber} ({chunkStartPosition} => {chunkEndPosition})",
+				oldChunkName,
+				physicalWeight,
+				chunkStartNumber, chunkEndNumber, chunkStartPos, chunkEndPos);
+
+			IChunkWriterForExecutor<TStreamId, TRecord> outputChunk;
+			try {
+				outputChunk = _chunkManager.CreateChunkWriter(sourceChunk);
+				Log.Trace(
+					"SCAVENGING: Resulting temp chunk file: {tmpChunkPath}.", 
+					Path.GetFileName(outputChunk.FileName));
+
+			} catch (IOException ex) {
+				Log.ErrorException(ex,
+					"IOException during creating new chunk for scavenging purposes. " +
+					"Stopping scavenging process...");
+				throw;
```

Comment:
same question here (https://github.com/EventStore/EventStore/pull/3494/files#r913413407) and in other places in ExecutePhysicalChunk() that throw an exception